src/SFML/Graphics/VertexBuffer.cpp

```diff
@@ -0,0 +1,373 @@
+
+    GLboolean destinationResult = GL_FALSE;
+    glCheck(destinationResult = GLEXT_glUnmapBuffer(GLEXT_GL_ARRAY_BUFFER));
+
+    glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, 0));
+
+    if ((sourceResult == GL_FALSE) || (destinationResult == GL_FALSE))
+        return false;
+
+    return true;
+
+#endif // SFML_OPENGL_ES
+}
+
+
+////////////////////////////////////////////////////////////
+void VertexBuffer::clear()
+{
+    if (m_buffer)
+    {
+        TransientContextLock contextLock;
+
+        glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, m_buffer));
+        glCheck(GLEXT_glBufferData(GLEXT_GL_ARRAY_BUFFER, 0, 0, usageToGlEnum(m_usage)));
+        glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, 0));
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+VertexBuffer& VertexBuffer::operator =(const VertexBuffer& right)
+{
+    VertexBuffer temp(right);
+
+    swap(temp);
+
+    return *this;
+}
+
+
+////////////////////////////////////////////////////////////
+void VertexBuffer::swap(VertexBuffer& right)
+{
+    std::swap(m_size,          right.m_size);
+    std::swap(m_buffer,        right.m_buffer);
+    std::swap(m_primitiveType, right.m_primitiveType);
+}
+
+
+////////////////////////////////////////////////////////////
+unsigned int VertexBuffer::getNativeHandle() const
+{
+    return m_buffer;
+}
+
+
+////////////////////////////////////////////////////////////
+void VertexBuffer::bind(const VertexBuffer* vertexBuffer)
+{
+    if (!isAvailable())
+        return;
+
+    TransientContextLock lock;
+
+    glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, vertexBuffer ? vertexBuffer->m_buffer : 0));
+}
+
+
+////////////////////////////////////////////////////////////
+void VertexBuffer::setPrimitiveType(PrimitiveType type)
+{
+    m_primitiveType = type;
+}
+
+
+////////////////////////////////////////////////////////////
+PrimitiveType VertexBuffer::getPrimitiveType() const
+{
+    return m_primitiveType;
+}
+
+
+////////////////////////////////////////////////////////////
+void VertexBuffer::setUsage(VertexBuffer::Usage usage)
+{
+    m_usage = usage;
+}
+
+
+////////////////////////////////////////////////////////////
+VertexBuffer::Usage VertexBuffer::getUsage() const
+{
+    return m_usage;
+}
+
+
+////////////////////////////////////////////////////////////
+bool VertexBuffer::isAvailable()
+{
+    Lock lock(isAvailableMutex);
+
+    static bool checked = false;
+    static bool available = false;
+
+    if (!checked)
+    {
+        checked = true;
+
+        TransientContextLock contextLock;
+
+        // Make sure that extensions are initialized
+        sf::priv::ensureExtensionsInit();
+
+        available = GLEXT_vertex_buffer_object;
+    }
+
+    return available;
+}
+
+
+////////////////////////////////////////////////////////////
+void VertexBuffer::draw(RenderTarget&, RenderStates) const
+{
+    // This should never be called
```

Comment:
Why is it a `Drawable` then?

Reply:
The Drawable API until now has been built around the idea that everything is representable and drawable through a client side array. While sf::VertexBuffer is also conceptually a Drawable, it can't make use of this API because there is no client side pointer to pass to sf::RenderTarget. Maybe the Drawable API has to be extended/changed but this would be too big of a breaking change for SFML 2.

Reply:
> While sf::VertexBuffer is also conceptually a Drawable

I'd say it is drawable, but it is not a `sf::Drawable`. That would translate to: it can be passed to `RenderTarget::draw`, which is exactly what you've done with the new overload.

A simple and probably common use case where making it a derived `sf::Drawable` fails, is a container of `sf::Drawable` instances and drawing them all from this container. That would call the `draw(sf::Drawable&)` overload, which would call `VertexBuffer::draw`, ending up in nothing at all being done.

At least, make `VertexBuffer::draw` call `draw(sf::VertexBuffer&)`, like `sf::VertexArray` does.

Reply:
Wouldn't this end up in infinite recursion?

RenderTarget::draw() -> VertexBuffer::draw() -> RenderTarget::draw() -> VertexBuffer::draw() ...

Reply:
No, because once you're in VertexBuffer::draw, you can pass a VertexBuffer& (*this) instead of a Drawable&, so it ends up in the correct overload and draws the vertex buffer.

Reply:
Yes... but the user's call into RenderTarget::draw() would already be using the VertexBuffer& overload instead of the Drawable& one...

Reply:
That's why I mentioned the case where the user draws the vertex buffer instance through its base class.

```cpp
std::vector<sf::Drawable*> drawables;
drawable.push_back(new sf::VertexBuffer);
...
for (auto drawable : drawables)
    window.draw(*drawable); // calls RenderTarget::draw(sf::Drawable&)
```